### PR TITLE
Add `scanMap`

### PR DESCRIPTION
**Pull request by inamiy**
Hi all ✋ 

This PR will add a new `Signal.transduce` feature that is:

- More functionality than `scan`
- Essentially, it is `traverse :: Applicative f => (a -> f b) -> t a -> f (t b)` where `f = State s = (s -> (output, s))` and `t = Signal`
- `transduce` works similar to [Mealy Machine](https://en.wikipedia.org/wiki/Mealy_machine), or [Elm](https://elm-lang.org/)'s `update` where `Output` is `SignalProducer` (effect).

Furthermore:

- The name `transduce` comes from Mealy machine that is "deterministic finite-state transducer".
- This is a simplified version of [ReactiveAutomaton](https://github.com/inamiy/ReactiveAutomaton) with completely hidden `State`.
- Reducing with function `(State, Value) -> (State, Output)` is powerful as presented by @mbrandonw in [Functional Swift Conference 2017](https://www.youtube.com/watch?v=QOIigosUNGU&feature=youtu.be&t=2608)

I would like to hear your thoughts on this ✨ 

## TODO

- [x] Add comment
- [x] Updated CHANGELOG.md.

### CODE REVIEW

**Comment by andersio** [comment]
Looks nice! Could we have this documented please?

**Comment by inamiy** [comment]
@andersio 
Thanks! Updated doc in 1058068.

**Comment by inamiy** [comment]
### Alternative names to consider
	
- `mapAccumulate` / `accumulate`
    - originally from `mapAccumR :: Traversable t => (a -> b -> (a, c)) -> a -> t b -> (a, t c)`

- `traverse` 
    - originally from `traverse :: Applicative f => (a -> f b) -> t a -> f (t b)`

- `mealy`
	- originally from [Mealy Machine](https://en.wikipedia.org/wiki/Mealy_machine), or [`Data.Machine.Mealy unfoldMealy`](http://hackage.haskell.org/package/machines-0.6.4/docs/Data-Machine-Mealy.html) has similarity

**Comment by andersio** [comment]
Maybe `scanMap`? There has no precedence but the operation is closely related to `scan`. 🤔

cc @ReactiveCocoa/reactiveswift

**Comment by andersio** [comment]
Can we also see if `scan` can be implemented with this operator by composing the transform?

**Comment by inamiy** [comment]
> Can we also see if scan can be implemented with this operator by composing the transform?

Done in 6bca910 !

Also renamed the method to `scanMap` in 6a465c3 .
Now two methods look more close to each other.

**Comment by mdiep** [comment]
Would you mind giving an example of what you'd use this for?

**Comment by inamiy** [comment]
In my case, I want to map video stream's `metadata` to also emit `ExtraFlag` based on first "playing metadata" + that might get reset via `metadata = nil`:

```swift
typealias State = Metadata? // PlayingMetadata
typealias ExtraFlag = Bool

let upstream: Signal<Metadata?, NoError> = ...

let downstream: Signal<(Metadata, ExtraFlag)?, NoError> =
    upstream.scanMap(nil) { (state: State?, metadata: Metadata?) -> (State?, (Metadata, ExtraFlag)?) in

        guard let metadata = metadata else {
            // `metadata = nil` as resetting state
            return (nil /* reset state */ , nil /* no metadata nor extra flag */ )
        }
       
        guard let state = state else {
            // 1st `metadata` as `state` if `metadata.isPlaying`, otherwise don't store state & send
            return metadata.isPlaying ? (metadata, (metadata, false)) : (nil, nil)
        }
        
        // Non-first `metadata` comes here
        switch state {
            case .playingMetadata1:
                return (state /* no change */, (metadata, false))
            case .playingMetadata2:
                return (state /* no change */,  (metadata, true)) // special flag attached
        }

    }
```

In this complicated scenario, `state` is often tweaked via `upstream`'s value, but emits different output than `Metadata?` that depends on internal state.

### P.S.

BTW, you can rewrite `scanMap` with `scan`, but `State` becomes messy and not efficient:

```swift
// RxSwift example
public func scanMap<State, Output>(
    into initial: State,
    _ f: @escaping (inout State, E) -> Output
    ) -> Observable<Output> {

    return self
        .scan(into: (initial, Optional<Output>.none), accumulator: { tuple, element in
            let output = f(&tuple.0, element)
            tuple.1 = output
        })
        .flatMap { tuple -> Observable<Output> in
            if let output = tuple.1 {
                return .just(output)
            } else {
                return .empty()
            }
        }
}
```

**Comment by inamiy** [comment]
ff3851a https://travis-ci.org/ReactiveCocoa/ReactiveSwift/builds/469555480
CI build error seems unrelated.

**Comment by andersio** [comment]
In our codebase there is a common scenario to keep the first value and drop subsequent values until `nil` occurs, which currently implements as a `scan` plus a `filterMap`:

```swift
extension SignalProducer {
    func firstUntilNil<U>(_ transform: @escaping (Value) -> U?) -> SignalProducer<U, Error> {
        return self
            .scan(into: (true, nil)) { (current: inout (isPreviouslyNil: Bool, lastOutput: U?), newValue: Value) in
                switch transform(newValue) {
                case let .some(value) where current.isPreviouslyNil:
                    current = (isPreviouslyNil: false, lastOutput: value)
                case .some:
                    current = (isPreviouslyNil: false, lastOutput: nil)
                case .none:
                    current = (isPreviouslyNil: true, lastOutput: nil)
                }
            }
            .filterMap { $0.lastOutput }
    }
}
```

which could be made better by `scanMap` in the sense that the operator no longer needs to keep a copy of the output just to make the compiler & the operator happy:
```swift
extension SignalProducer {
    func firstUntilNil<U>(_ transform: @escaping (Value) -> U?) -> SignalProducer<U, Error> {
        return self
            .scanMap(into: true) { (isPreviouslyNil: inout Bool, newValue: Value) -> U? in
                switch transform(newValue) {
                case let .some(value) where isPreviouslyNil:
                    isPreviouslyNil = false
                    return value
                case .some:
                    return nil
                case .none:
                    isPreviouslyNil = true
                    return nil
                }
            }
            .skipNil()
    }
}
```

**Comment by mdiep** [comment]
I'm not convinced that this is a worthwhile addition.

> In my case, I want to map video stream's `metadata` to also emit `ExtraFlag` based on first "playing metadata" + that might get reset via `metadata = nil`:

It seems like you'd be better off adding a `struct` instead of a `tuple` and handling the mutation there.

```swift
struct State {
  var metadata: Metadata?
  var extraFlag: Bool
}

let upstream: Signal<Metadata?, NoError> = ...

let downstream: Signal<(Metadata, ExtraFlag)?, NoError> =
    upstream.scanMap(nil) { (state: State?, metadata: Metadata?) -> State in
        guard let metadata = metadata else {
            // `metadata = nil` as resetting state
            return nil
        }
       
        guard let state = state else {
            // 1st `metadata` as `state` if `metadata.isPlaying`, otherwise don't store state & send
            return metadata.isPlaying ? State(metadata: metadata, extraFlag: false) : nil
        }
        
        // Non-first `metadata` comes here
        var newState = state
        newState.metadata = metadata // this would update the extra flag
        return newState
    }
```

> In our codebase there is a common scenario to keep the first value and drop subsequent values until `nil` occurs, which currently implements as a `scan` plus a `filterMap`:

I think this would be more clearly written with `combinePrevious`, since that's what you really care about:

```swift
extension SignalProducer {
    func firstUntilNil<U>(_ transform: @escaping (Value) -> U?) -> SignalProducer<U, Error> {
        return self
            .map(transform)
            .combinePrevious(initial: nil)
            .flatMap(.concat) { ((old, new)) in
              return old == nil ? .init(value: new) : .empty
            }
            .skipNil()
    }
}
```

**Comment by andersio** [comment]
> I think this would be more clearly written with `combinePrevious`, since that's what you really care about

It would have to hold onto the previous value, even if expressed using `skipRepeats`, which might not be desirable sometimes. The operator here is apparently also a more general form which some other operators like `combinePrevious` and `skipRepeats` can be built on top of, now with only a transform.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 1. New operator `materializeResults` and `dematerializeResults` (#679, kudos to @ra1028)
 1. New convenience initializer for `Action` that takes a `ValidatingProperty` as its state (#637, kudos to @Marcocanc)
 1. Fix legacy date implementation. (#683, kudos to @shoheiyokoyama)
+1. New operator `transduce`. (#695, kudos to @inamiy)
```
[the remark this answers is not in the document]
Comment (inamiy) [reply CHANGELOG.md]: Done in cbca7339.

File: Sources/Signal.swift [hunk 2]
```diff
@@ -1314,6 +1314,34 @@ extension Signal {
 	public func scan<U>(into initialResult: U, _ nextPartialResult: @escaping (inout U, Value) -> Void) -> Signal<U, Error> {
 		return flatMapEvent(Signal.Event.scan(into: initialResult, nextPartialResult))
 	}
+
+	/// Accumulate all values from `self` as `State`, and send the value as `Output`.
```
[the remark this answers is not in the document]
Comment (inamiy) [reply Sources/Signal.swift]: Done in cbca7339.

File: Sources/Signal.swift [hunk 3]
```diff
@@ -1314,6 +1314,34 @@ extension Signal {
...
+	///           from `self`. The result would be "next state" and "output" where
+	///           "output" would be forwarded and "next state" would be used in the
+	///           next call of `next`.
+	///
+	/// - returns: A producer that sends the output that is computed from the accumuation.
+	public func scanMap<State, U>(_ initialState: State, _ next: @escaping (State, Value) -> (State, U)) -> Signal<U, Error> {
+		return flatMapEvent(Signal.Event.scanMap(initialState, next))
+	}
+
+	/// Accumulate all values from `self` as `State`, and send the value as `Output`.
```
[the remark this answers is not in the document]
Comment (inamiy) [reply Sources/Signal.swift]: Done in cbca7339.

**Comment by inamiy** [comment]
> I'd expect an operator to hold just the minimal viable state for doing its business.

Yes! This is the whole goal of making this simplest API by clearly separating the incoming value, internal state, and outgoing value.

To simplify more, this discussion is no more than:

```swift
.scan((state, nil)) { /* complicated code because of caching output */ }
.filterMap { $0.1 }
// or any caching operator + flatMap
```

V.S.

```swift
.scanMap(state) { /* simpler code */ }
```

as mentioned earlier, and it's verbose to implement the former code.

Though readability might degrade a little by using this operator only, I think it is much better than caching unused value.

**Comment by inamiy** [comment]
Updated to track current master branch.

@andersio Can you review this PR?

**Pull request merged**